MSC2881: Message Attachments

MurzNN opened the pull request:
[Rendered](https://github.com/MurzNN/matrix-doc/blob/master/proposals/2881-message-attachments.md)

Depends on [MSC3051: A scalable relation format](https://github.com/matrix-org/matrix-doc/pull/3051)

Similar MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/4274

Status: Closed in favor of inline media references like this:
```
Check out my photos from FOSDEM 2025:
<img src=\"mxc://example.org/JWEIFJweifjWIeifj\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifk\" />" 
<img src=\"mxc://example.org/JWEIFJweifjWIeifl\" />" 

And especially look at this awesome group photo: 
<img src=\"mxc://example.org/JWEIFJweifjWIeifm\" />" 

What a nice day at FOSDEM 2025!
```
_To implement this feature, we don't need any server-side changes; this can be implemented fully on the client side (message composer to handle inserting, and renderer to display thumbnails and pop-up galleries). 

This idea is described in a separate MSC: https://github.com/matrix-org/matrix-spec-proposals/pull/4475

Related issues:
- [Attach multiple images (files) to message](https://github.com/vector-im/element-meta/issues/685)
- [Show albums for multiple images](https://github.com/vector-im/element-meta/issues/753)
- [Message attachments (multiple photos, files, other messages for forwarding, etc)](https://github.com/matrix-org/matrix-doc/issues/2289)
- [Redesign: timeline: implement new file transfer UX](https://github.com/vector-im/element-web/issues/7565
- https://github.com/vector-im/roadmap/issues/13

Also, look up alternative and more optimal implementation: matrix-org/matrix-spec-proposals#3382

Signed-off-by: Alexey Korepov <MurzNN@gmail.com>

MurzNN committed:
Added proposial "Message Attachments"

MurzNN committed:
Assign 2881 number to MSC and some typos

deepbluev7 commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,123 @@
+<div class="mx-attachments">
+  <p>Attachments:</p>
+  <ul>
+    <li><a href="https://example.com/_matrix/media/r0/download/example.com/KUAQOesGECkQTgdtedkftISg">Image 1.jpg</a></li>
```
The fallback should probably use mxc uris, as not every media repo allows downloads without authentication and it puts a lot more strain on a server, if you post in matrix.org and everyone uses the fallback instead of letting matrix.org act as a media proxy.

MurzNN replied:
In my first draft was exactly mxc urls, but I observe, that they don't work in current Element-web (and in most of other clients), so replace them to http urls. If mxc urls must work by spec, then yes - we can replace this to mxc urls.

deepbluev7 replied:
Fair enough. mxc in `a` tags is still an open MSC: https://github.com/matrix-org/matrix-doc/pull/2398

Maybe mention that consideration in the proposal?

MurzNN replied:
Yes, wasn't seen it, done now!

uhoreg replied:
I'm surprised that an http(s) url even works (and it probably is a bug in the client if it does).  Generally clients try to avoid making requests to things other than their homeservers, to avoid leaking the user's IP address to an unknown party.

deepbluev7 replied:
I don't think it is a bug to allow the user to click on links. It would be a different issue, if the spec allowed http(s) as a valid protocol for inline images.

uhoreg replied:
Ah, right, it's an `<a>`, not an `<img>`.  Disregard my comment, then.

deepbluev7 commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,123 @@
+
+## Alternatives
+
+1. An alternative can be posting media messages as separate events, as it was done earlier, and aggregating them into event via  `m.relates_to` field, but clients must do a hide of those events, when aggregating event will be added to the room (like editions), but this will be harder to implement.
```
```suggestion
1. An alternative can be posting media messages as separate events, as it was done earlier, and aggregating them into event via  `m.relates_to` field, but clients must do a hide of those events, when aggregating event will be added to the room (like edits), but this will be harder to implement.
```

Also I disagree, that this would be any harder. Only issue may be the send order, but if you send the text message first and then the attachments after it, that should be fine. Especially if you do all the uploads before sending the first event, this should be fine. You may also add a flag to the first event, that there will be attachments, so rendering can reserve some space.

But that would give you fallbacks for free, work with existing bridges and clients and sidestep all the issues the fallback introduces. I think Whatsapp does something similar, when it groups a big drop of images into one message (just without explicit aggregations, I think).

MurzNN replied:
Interesting idea, thanks! For fallback this way is better, but it will generate many useless events in room, instead of one grouped event. I'll try to describe this way too, but still can't decide which implementation is better... How to better describe this implementation - add section "Variant 2" in current MSC, or post a separate MSC with this implementation?

P. S. About WhatsApp - yes, do the grouping on client side, if see several events with images from one user in short time, but I didn't like this way, because of missing real grouping event and attached text description.

deepbluev7 replied:
>  How to better describe this implementation - add section "Variant 2" in current MSC, or post a separate MSC with this implementation?

That is actually a good question. Since this MSC is fairly small, I'd just put both into this MSC by splitting the `Proposal` section into a short introduction and then 2 subheading with the 2 different variants, while naming each of those. A separate MSC is fine too, but I think as this has a clear defined scope and the implementation part is somewhat short in both cases, discussing both here should be fine. I'll let someone more experienced cast their judgement though.

deepbluev7 replied:
I also don't see the event count as a concern. Every reaction adds an event and you don't save much bandwidth by sending everything in one event. You can also have the server aggregate everything into one event in the future, if you like, to alleviate the rendering concerns.

MurzNN replied:
I added this way as "Implementation 2: Aggregating event to group several previously sent media events" section.

MurzNN replied:
If event count is not a concern - yes, this is better implementation that my first, so I swapped them now ;-)

deepbluev7 commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,123 @@
+1. An alternative can be posting media messages as separate events, as it was done earlier, and aggregating them into event via  `m.relates_to` field, but clients must do a hide of those events, when aggregating event will be added to the room (like editions), but this will be harder to implement.
+2. Other alternative is embedding images (and other media types) into message body via html tags, but this will make extracting and stylizing of the attachments harder.
+3. Next alternative is reuse [MSC1767: Extensible events in Matrix](https://github.com/matrix-org/matrix-doc/pull/1767) for attaching and grouping media attachments, but in current state it requires only one unique type of content per message, so we can't attach, for example, two `m.image` items into one message. Maybe, instead of separate current issue, we can extend [MSC1767](https://github.com ...
+
```
You may want to mention https://github.com/matrix-org/matrix-doc/pull/2530 and https://github.com/matrix-org/matrix-doc/pull/2529 as competing proposals and a justification, how this improves upon them (and potential drawbacks compared to them), like that this allows sending multiple images in one message.

MurzNN replied:
Yes, thanks for pointing on them, I add both to alternatives section.

sumnerevans replied:
I don't think that these proposals are necessarily mutually exclusive. There are benefits to having captions tied more directly to a single image, and then having an aggregation message that allows for multiple images to be combined into a gallery.

MurzNN committed:
Added mentions of similar MSC

#2530 and #2529

MurzNN committed:
Direct mxc links and improved alternative

deepbluev7 commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,129 @@
+3. Next alternative is reuse [MSC1767: Extensible events in Matrix](https://github.com/matrix-org/matrix-doc/pull/1767) for attaching and grouping media attachments, but in current state it requires only one unique type of content per message, so we can't attach, for example, two `m.image` items into one message. Maybe, instead of separate current issue, we can extend [MSC1767](https://github.com ...
+
+4. There are also [MSC2530: Body field as media caption](https://github.com/matrix-org/matrix-doc/pull/2530) but it describes only text description for one media, not several media items, and very similar [MSC2529: Proposal to use existing events as captions for images](https://github.com/matrix-org/matrix-doc/pull/2529) that implement same thing, but via separate event. But if we send several me ...
+
```
I think this MSC is missing an unstable prefix: https://github.com/matrix-org/matrix-doc/pull/2371

At least `m.attachments` and `mx-attachments` would require it, imo.

MurzNN replied:
Done

MurzNN committed:
Improved description of 1'st alternative

MurzNN committed:
Added second implementation

MurzNN committed:
Added unstable prefixes, fixed typo

MurzNN committed:
Added unstable to class to

MurzNN committed:
Moved fallback dispay to implementation 1

MurzNN committed:
Added example of media event to implentation 2

MurzNN committed:
Move second implementation to main place

MurzNN committed:
Paraphrasing

sumnerevans commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,164 @@
+  "type": "m.room.message",
+  "content": {
+    "msgtype": "m.text",
+    "body": "Here is my photos and videos from yesterday event",
```
I think this would make it difficult to support a gallery where each of the photos has its own caption with this system. Maybe it's worth implementing both this and either #2530 or #2529 and use this for galleries, but use the the other proposal to handle captions?

MurzNN replied:
Message with attachments can work together with individual captions, like proposed in #2530 and #2529, because it simply refers to other event, that can have any structure. But yes - gallery, where each image have own caption, is too complex for simple chatroom, and will be not so easy to implement in UI, especially on mobiles. So I vote to keep this simple: one text message + several attachments.

MurzNN replied:
I have added description of combining with #2530 into description of MSC.

sumnerevans commented on proposals/2881-message-attachments.md:
````diff
@@ -0,0 +1,164 @@
+}
+```
+
+## Client support
````
We should probably also consider how editing should work. I'd imagine the following scenarios:

* Add caption: Add the "edit" pencil-button to pictures, and if the user clicks on that, return them to the same UI as the composer allowing them to add/remove files or edit the caption.
* Edit caption: Same flow as adding a caption.
* Remove caption: Add a "remove caption" button to the dialog if editing.
* Remove attachment: Allow attachments to be removed from the gallery with the same UI as when composing the message. One question here would be what to do with the corresponding image (or other attachment) event, probably just delete it?
* Add attachment: Allow attachments to be added to the gallery with the same UI as when composing the message. This should create a new `m.image` event and then edit the aggregating message.

MurzNN replied:
As I see, editing can be represented exactly like composer interface, where user have the textarea for input message text, and area with all current attachments as tiny thumbnails, in which he can remove one of current attachments (that will remove its line from array of `m.relates_to`), add new attachment (that will upload it as new event, and refer to it in edited message).

And refer to edited message will be kept as another item of array, with `"rel_type": "m.replace"`.

sumnerevans replied:
On remove, should the corresponding `m.image` event also be removed? That way, clients trying to derive the state in the future will not be confused and accidentally show that image (even though it was removed from the gallery).

MurzNN replied:
Yes, it can (maybe even should) be redacted, like current 'redact` action on m.file events. I add this to description, thanks for pointing!

MurzNN committed:
Described edits and improved description

MurzNN committed:
Added fallback description

MurzNN committed:
Added edits composer description

MurzNN committed:
Added link to extensible events suggestion

MurzNN committed:
Added alternative with attachments after message

MurzNN committed:
Added redact action on remove media

MurzNN committed:
Fixed typo

MurzNN committed:
Fix typo again

MurzNN committed:
Added delete (redact action)

MurzNN committed:
Added potential issue and hiding recommendation

MurzNN committed:
Added link to MSC2278

MurzNN committed:
Fix typos and added link to MSC2675

MurzNN committed:
Fix typo

MurzNN committed:
Added proposal for replacement to #2530 / #2529

MurzNN committed:
Added links to implementation examples

MurzNN committed:
Improved description

MurzNN committed:
Description for implementation together with #2530

erkinalp commented on proposals/2881-message-attachments.md:
````diff
@@ -0,0 +1,239 @@
+  "url": "mxc://example.com/KUAQOesGECkQTgdtedkftISg"
+},
+```      
+But this way will give harder way to render of main message event, because Matrix clients must do the search of all attached events manually in timeline, and server will can't aggregate them to main message.
````
```suggestion
But this way will give harder way to render of main message event, because Matrix clients must do the search of all attached events manually in timeline, and server will be unable to aggregate them to main message.
```

MurzNN committed:
improve grammar

Co-authored-by: Erkin Alp Güney <erkinalp9035@gmail.com>

MurzNN committed:
Move back first alternative as option two

MurzNN committed:
Fix typo

MurzNN committed:
Described replacing fallback to rich representation

MurzNN committed:
Name options as primary-secondary

MurzNN commented:
The [Hummingbard](https://hummingbard.com/) project seems use "Option two" implementation from this MSC to attach multiple files/images/links to event! 🎉

MurzNN committed:
Named options

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+*This MSC is especially for media image attachments to message, but I try to make it extendable for multiple attachment types (files, videos, and in future - external URLs, links to other Matrix events, etc). So, in most of examples, I am using "image", but it means, that, instead of image, there may be another attachment type.*
+
+In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to his text message when he is composing it,  not after or before it.
```
```suggestion
In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to their text message when composing it, not after or before it.
```

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+In the current implementation each media (file, image, video) can be sent only via a separate event to the room. But in most cases even one media is not sent alone, it must be commented via some text by the sender. So the user often wants to attach some images (one or several) directly to his text message when he is composing it,  not after or before it.
+
+And now the user can send images only before the message (or after it) as a separate message, but he can't attach images during the composing process to send them when the text is finished, together with the text message in one event.
```
```suggestion
And now the user can send images only before the message (or after it) as a separate message, but they can't attach images during the composing process to send them when the text is finished, together with the text message in one event.
```

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+When user press "Send" button, Matrix client do the upload of all media, that user attached to message, as separate events to room (how it is done now), before sending message with typed text. And after sending of all attachments is finished, client send message with aggregating event, using `m.relates_to` field (from the [MSC2674: Event relationships](https://github.com/matrix-org/matrix-doc/pul ...
+
+For exclude showing those events in modern clients before grouping event added, I propose extend separate media events via adding "marker" field `is_attachment: true`, if clients got this value - they must exclude showing this media in timeline, and shows them only in gallery with grouping event.
```
What happens if such an event is never referenced? Does it just stay hidden forever? This is probably the best option but it is worth discussing.

Pros:
- Allows uploading the media as the user composes so that sending is quicker.
- Messages don't suddenly appear (for example) a day later higher up in the timeline.
- If the sender crashes or fails it can just restart. It isn't *required* to recollect the event IDs and reuse them. (Although reuse would be more efficient)

Cons:
- Invisible events may be an abuse vector. (DoS or covert channel)
- Upon a failed send or deleted message (if the client eagerly uploads media) the user likely doesn't realize that the media was uploaded and available in the room history. This may be a concern if the user then realized that they didn't want to send that info. (Maybe they tried to cancel the send because it was in the wrong room and would have redacted the media if they realized it was actually sent).

So basically it is nice because it emulates an atomic commit. But it is not nice because it does expose some side effects even if not committed.

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+### Display recommendations:
+
+On the client site, attachments can be displayed as grid of clickable thumbnails, like the current `m.image` events, but with a smaller size, having fixed height, like a regular image gallery. On click, Matrix client must display media in full size, and, if possible, as a gallery with "next-previous" buttons. Also clients can implement collapse/expand action on gallery grid.
```
For a large number of attachments (for example "look at thses photos from my two week vacation!") many clients won't want to show all of the images upfront. Most would probably render some sort of grid with the first couple then a "37 more" link.

In order to help with cross-client consistency should we specify something about the order and how that should be handled? You do mention "rearranging" attachments earlier.

For example:
- Clients SHOULD show the attachments in the order the appear in the `m.relates_to` array.
- If the client decides to show only a subset of photos upfront it SHOULD prefer to show the ones earliest in the `m.relates_to` array.

I think it should be non-normative, but basically if you are going to only show some show the first ones rather than the last ones.

kevincox replied:
I don't think we need to specify "must" here. For example if the image is bigger than the users display I think most users would rather have it scaled to fit rather than needing to scroll around. I would leave this section as an example of a possible implementation.

kevincox replied:
```suggestion
On the client side, attachments can be displayed as grid of clickable thumbnails, like the current `m.image` events, but with a smaller size, having fixed height, like a regular image gallery. On click, Matrix client must display media in full size, and, if possible, as a gallery with "next-previous" buttons. Also clients can implement collapse/expand action on gallery grid.
```

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+## Potential issues
+
+1. On bad connection to server Matrix client can send attachments as events with `"is_attachment": true` but not send final `m.message` event, this will lead to posting invisible media to room. This can be solved on client side via caching unsent group of events, and repeat sending when connection will be recovered.
```
What happens if such an event is never referenced? Does it just stay hidden forever? This is probably the best option but it is worth discussing.

Pros:
- Allows uploading the media as the user composes so that sending is quicker.
- Messages don't suddenly appear (for example) a day later higher up in the timeline.
- If the sender crashes or fails it can just restart. It isn't *required* to recollect the event IDs and reuse them. (Although reuse would be more efficient)

Cons:
- Invisible events may be an abuse vector. (DoS or covert channel)
- Upon a failed send or deleted message (if the client eagerly uploads media) the user likely doesn't realize that the media was uploaded and available in the room history. This may be a concern if the user then realizes that they didn't want to send that info. (Maybe they tried to cancel the send because it was in the wrong room and would have redacted the media if they realized it was actually sent).

So basically it is nice because it emulates an atomic commit. But it is not nice because it does expose some side effects even if not committed.

pavlukivan replied:
there is already many ways to send an invisible event, that by itself shouldn't be considered an issue. The other points are still valid though

kevincox commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+1. On bad connection to server Matrix client can send attachments as events with `"is_attachment": true` but not send final `m.message` event, this will lead to posting invisible media to room. This can be solved on client side via caching unsent group of events, and repeat sending when connection will be recovered.
+
+2. In option one - individual media event, to which `m.message` refers, can be deleted (redacted) after. As result, `m.message` will contain relation to redacted event. In this situation Matrix clients can exclude this item from display.
```
I think this should be put into the MSC proper. I think it makes sense that redacting an attachment implies that the attachment is removed from the messages that include it. However I would leave it open to clients showing a "redacted" placeholder or just hiding it outright. (For example if an attachment was referenced from multiple events it may be helpful to see which events now have missing attachments).

MurzNN committed:
Grammar fixes

Co-authored-by: Kevin Cox <kevincox@kevincox.ca>

MurzNN committed:
Grammar fixes

Co-authored-by: Kevin Cox <kevincox@kevincox.ca>

MurzNN committed:
Grammar fixes

Co-authored-by: Kevin Cox <kevincox@kevincox.ca>

kevincox commented:
This seems like a clear answer of "we will need it". It is easy to imagine and we have a real-world use case here.

pavlukivan commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+
+With having this feature, Matrix client should allow users to attach one or multiple media (images, video, files) to message on client side, without instant sending of them to room, and send them together with text message.
+
+When user press "Send" button, Matrix client do the upload of all media, that user attached to message, as separate events to room (how it is done now), before sending message with typed text. And after sending of all attachments is finished, client send message with aggregating event, using `m.relates_to` field (from the [MSC2674: Event relationships](https://github.com/matrix-org/matrix-doc/pul ...
```
Can you change it to use https://github.com/matrix-org/matrix-doc/pull/3051 (and thus, m.relations instead of m.relates_to)? MSC2674 only supports a single relation per message.

MurzNN replied:
I added link to #3051 additionally to #2674, to keep both of them in sight.

pavlukivan commented:
> But so far there seem to be not enough use cases to allow multiple relations.

Off the top of my head: replying to multiple message at the same time could be done with that

MurzNN committed:
Merge branch 'matrix-org:main' into master

MurzNN committed:
Link to MSC3051 and separate alternative version to 3382

Option two is moved to https://github.com/matrix-org/matrix-doc/pull/3382
Also added link to https://github.com/matrix-org/matrix-doc/pull/3051 with array implementation of relations.

MurzNN commented on proposals/2881-message-attachments.md:
```diff
@@ -0,0 +1,236 @@
+Messages with multiple attachments now already implemented in many messengers, for example - in Skype, Slack, VK Messenger. And Matrix, because lack of support, now have problems with bridging those messages to Matrix rooms.
+
+
+## Proposal
```
I moved option two to separate #3382

turt2live commented:
@MurzNN I'm not sure what happened here, but the diff is showing quite a few changes. I would recommend either fixing the branch, or, if preferred, closing this PR and opening a new MSC off a dedicated branch.

richvdh commented:
suspect this is the problem:

![image](https://user-images.githubusercontent.com/1389908/143013876-c69575f3-e72a-4e08-b58d-cf20c07ac94e.png)

see https://github.com/matrix-org/matrix-doc/issues/3367

MurzNN commented:
@richvdh Thanks for the tip, I've switched to `main` branch, seems now all is well.
